### ISSUE TITLE
Curve cannot handle decorated functions

### ISSUE THREAD

**Issue by hoogerheide** [bumps/bumps#65]
Using a decorated function (e.g. jit) as the input to Curve generates an "unsupported callable" error. This occurs because Curve.__init__ calls inspect.getfullargspec which cannot handle decorated functions either. A potential workaround for inspect is in this thread: https://github.com/lmfit/lmfit-py/issues/570#issuecomment-516637403, but the overall recommendation appears to be to move to inspect.signature instead.

Note that calling the jit-ted function with another function does not give this error (e.g. def fitfunc(x, m, b): return line(x, m, b)), but does produce a 25-50% performance hit, at least on my machine. Using lambda in the Curve definition also avoids this error, but I haven't characterized performance yet.

**Comment by bmaranville** [comment]
Can you post your affected model?

**Comment by hoogerheide** [comment]
Meant to attach with the original post. Here it is.
[curve_jit.py.txt](https://github.com/bumps/bumps/files/6564978/curve_jit.py.txt)

**Comment by bmaranville** [comment]
This replaces arginspect with signature: but now there is a new error, that says numba cannot compile line.  I throw it back over the fence to you, @hoogerheide 
```diff
diff --git a/bumps/curve.py b/bumps/curve.py
index 7457e77..72433cf 100644
--- a/bumps/curve.py
+++ b/bumps/curve.py
@@ -74,7 +74,14 @@ def _parse_pars(fn, init=None, skip=0, name=""):
     *pk=None* then the parameter will be non-fittable, and instead set
     through *init*.
     """
-    pnames, vararg, varkw, pvalues = inspect.getargspec(fn)
+    sig = inspect.signature(fn)
+    params = sig.parameters.values()
+    #pnames, vararg, varkw, pvalues = inspect.getfullargspec(fn)
+    pnames = [p.name for p in params]
+    pvalues = [p.default for p in params]
+    vararg = [p for p in params if p.kind.name == 'VAR_POSITIONAL']
+    varkw = [p for p in params if p.kind.name == 'VAR_KEYWORD']
+
     if vararg or varkw:
         raise TypeError(
             "Function %r cannot have *args or **kwargs in declaration"
@@ -88,11 +95,9 @@ def _parse_pars(fn, init=None, skip=0, name=""):
     defaults = dict((p, 0) for p in pnames)
 
     # If the function provides default values, use those.
-    if pvalues:
-        # Ignore default value for "x" parameter.
-        if len(pvalues) > len(pnames):
-            pvalues = pvalues[-len(pnames):]
-        defaults.update(zip(pnames[-len(pvalues):], pvalues))
+    for param in list(params)[skip:]:
+        if param.default is not inspect.Parameter.empty:
+            defaults[param.name] = param.default
 
     # Non-fittable parameters need to be sent in as None
     state_vars = set(p for p, v in defaults.items() if v is None)
```

**Comment by bmaranville** [comment]
Might be related to https://github.com/numba/numba/issues/4914 - "Numba expects the module from which the function is being compiled to exist in the namespace".

**Comment by bmaranville** [comment]
It looks like unspecifying __name__ in the loader context, i.e. in fitproblem.py
ctx = dict(__file__=filename, __package__=package) #, __name__=module)
allows the model to load.  I think __name__ = 'bumps.fitproblem' in this case, which is a module that exists in the execution namespace.  I wonder if we need to create an empty module in bumps that can serve as the importable name for the model?

**Comment by bmaranville** [comment]
This patch adds the module name to the DynamicPackage created in util.relative_import, which seems to solve the problem.

```patch
diff --git a/bumps/fitproblem.py b/bumps/fitproblem.py
index cbcc2e9..013cd32 100644
--- a/bumps/fitproblem.py
+++ b/bumps/fitproblem.py
@@ -782,8 +782,8 @@ def load_problem(filename, options=None):
     Raises ValueError if the script does not define problem.
     """
     # Allow relative imports from the bumps model
-    package = util.relative_import(filename)
     module = os.path.splitext(os.path.basename(filename))[0]
+    package = util.relative_import(filename, package=module)
 
     ctx = dict(__file__=filename, __package__=package, __name__=module)
     old_argv = sys.argv
diff --git a/bumps/util.py b/bumps/util.py
index 56a9741..36bdcd9 100644
--- a/bumps/util.py
+++ b/bumps/util.py
@@ -97,8 +97,9 @@ def kbhit():
 
 
 class DynamicPackage(object):
-    def __init__(self, path):
+    def __init__(self, path, name):
        self.__path__ = [path]
+       self.__name__ = name
 
 
 def relative_import(filename, package="relative_import"):
@@ -116,7 +117,7 @@ def relative_import(filename, package="relative_import"):
             and not isinstance(sys.modules[package], DynamicPackage)):
         raise ImportError("relative import would override the existing package %s. Use another name"
                           % package)
-    sys.modules[package] = DynamicPackage(path)
+    sys.modules[package] = DynamicPackage(path, package)
     return package
 
 

```

**Comment by bmaranville** [comment]
closed by #66
